### PR TITLE
feat(goal_planner): prioritize goals before objects to avoid

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

prioritize goals before objects to avoid.
gray numbers are the number of objects to avoid

object not to avoid

image

object to avoid

image

image

image

https://github.com/autowarefoundation/autoware.universe/assets/39142679/79af93ed-e17c-4d57-84d0-4be711837cce

## Related links

none
## Tests performed

psim

2023/09/27 https://evaluation.tier4.jp/evaluation/reports/d7c68b5d-3232-5881-8f18-71c2203f11b3/?project_id=prd_jt

## Interface changes

none

## Effects on system behavior

none

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kosuke55** [comment]
need to update path priority too.

```
```

image

**Comment by kosuke55** [comment]
https://github.com/autowarefoundation/autoware.universe/pull/5137#issuecomment-1735013792
fixed
image

File: planning/behavior_path_planner/include/behavior_path_planner/utils/goal_planner/goal_searcher.hpp [hunk 1]
```diff
@@ -36,14 +36,16 @@ class GoalSearcher : public GoalSearcherBase
     const GoalPlannerParameters & parameters, const LinearRing2d & vehicle_footprint,
     const std::shared_ptr<OccupancyGridBasedCollisionDetector> & occupancy_grid_map);
 
-  GoalCandidates search(const Pose & original_goal_pose) override;
+  GoalCandidates search() override;
   void update(GoalCandidates & goal_candidates) const override;
 
 private:
+  void countObjectsToAvoid(
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/include/behavior_path_planner/utils/goal_planner/goal_searcher.hpp:43]: [NITS] `countObjectsToPass` is more intuitive for me.

**Comment by kyoichi-sugahara** [comment]
smart!

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM!
launch change is not necessary?

**Comment by kosuke55** [comment]
@kyoichi-sugahara 
thanks! yes please https://github.com/autowarefoundation/autoware_launch/pull/594

**Pull request merged**